File: 2sum/optimized/solution.js
var twoSum = function (arr, target) {
    const map = new Map();
    
    for (let i = 0; i < arr.length; i++) {
        const rem = target - arr[i];
        if (map.has(rem)) {
            return [i, map.get(rem)];
        }
        map.set(i, arr[i]);
    }
  };
  
  console.log(twoSum([2, 7, 11, 15], 9));
  console.log(twoSum([3, 2, 4], 6));
  console.log(twoSum([3, 3], 6));
  

File: 2sum/optimized/solution.kt
import java.util.HashMap

internal object Solution {
    fun twoSum(arr: IntArray, target: Int): IntArray {
        val map = HashMap<Int, Int>()
        for (i in arr.indices) {
            val rem = target - arr[i]
            if (map.containsKey(rem)) {
                return intArrayOf(i, map[rem]!!)
            }
            map[arr[i]] = i
        }

        return intArrayOf()
    }

    @JvmStatic
    fun main(args: Array<String>) {
        val arr = intArrayOf(2, 7, 11, 15)
        val target = 9

        val result = twoSum(arr, target)
        println(result.contentToString())
    }
}